### PR TITLE
WIP: gated SAE impl

### PR DESCRIPTION
**Pull request by jettjaniak**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### You have tested formatting, typing and unit tests (acceptance tests not currently in use)

- [ ] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

### Performance Check.

If you have implemented a training change, please indicate precisely how performance changes with respect to the following metrics:
- [ ] L0
- [ ] CE Loss
- [ ] MSE Loss
- [ ] Feature Dashboard Interpretability

Please links to wandb dashboards with a control and test group.

### CODE REVIEW

File: sae_lens/sae.py [hunk 1]
```diff
@@ -235,19 +244,33 @@ def encode(
...
+            pre_acts_mag = torch.exp(self.r_mag) * z + self.b_mag
+            acts_mag = self.activation_fn(pre_acts_mag)
+            pre_acts_gate = z + b_gate
+            acts_gate = (pre_acts_gate > 0).float()
+            feature_acts = acts_gate * acts_mag
+        else:
+            # "... d_in, d_in d_sae -> ... d_sae",
+            hidden_pre = self.hook_sae_acts_pre(sae_in @ self.W_enc + self.b_enc)
+            feature_acts = self.activation_fn(hidden_pre)
+        self.hook_sae_acts_post(feature_acts)
```
Comment (jettjaniak) [inline sae_lens/sae.py:261]: why test failing

File: sae_lens/toolkit/pretrained_sae_loaders.py [hunk 2]
```diff
@@ -125,6 +125,8 @@ def connor_rob_hook_z_loader(
     cfg_dict = {
         "d_in": old_cfg_dict["act_size"],
         "d_sae": old_cfg_dict["dict_size"],
+        # TODO: is that OK?
```
Comment (jettjaniak) [inline sae_lens/toolkit/pretrained_sae_loaders.py:128]: it is OK

File: sae_lens/toolkit/pretrained_sae_loaders.py [hunk 3]
```diff
@@ -158,6 +160,10 @@ def load_pretrained_sae_lens_sae_components(
     cfg_dict["device"] = device
     cfg_dict["dtype"] = dtype
 
+    # TODO: is that OK?
+    if "architecture" not in cfg_dict:
```
Comment (jettjaniak) [inline sae_lens/toolkit/pretrained_sae_loaders.py:164]: yep

File: sae_lens/training/training_sae.py [hunk 4]
```diff
@@ -168,12 +171,28 @@ def encode_with_hidden_pre(
...
+            # TODO: pre-act hooks
+            b_gate = self.b_enc
+            # "... d_in, d_in d_sae -> ... d_sae"
+            z = sae_in @ self.W_enc
+            pre_acts_mag = torch.exp(self.r_mag) * z + self.b_mag
+            acts_mag = self.activation_fn(pre_acts_mag)
+            pre_acts_gate = z + b_gate
+            acts_gate = (pre_acts_gate > 0).float()
+            feature_acts = acts_gate * acts_mag
+            # FIXME: should we add noise at all? if so, to the gate or to the mag?
```
Comment (jettjaniak) [inline sae_lens/training/training_sae.py:184]: we should probably add noise to both or not allow noise in gated

File: sae_lens/training/training_sae.py [hunk 5]
```diff
@@ -168,12 +171,28 @@ def encode_with_hidden_pre(
...
+        else:
+            # "... d_in, d_in d_sae -> ... d_sae",
+            hidden_pre = self.hook_sae_acts_pre(sae_in @ self.W_enc + self.b_enc)
+            hidden_pre_noised = hidden_pre + (
+                torch.randn_like(hidden_pre) * self.cfg.noise_scale * self.training
+            )
+            feature_acts = self.activation_fn(hidden_pre_noised)
+        self.hook_sae_acts_post(feature_acts)
 
         return feature_acts, hidden_pre_noised
```
Comment (jettjaniak) [inline sae_lens/training/training_sae.py:197]: not great that `hidden_pre_noised` means different things in both branches

File: tests/unit/training/test_sae_trainer.py [hunk 6]
```diff
@@ -19,9 +19,9 @@
 from tests.unit.helpers import TINYSTORIES_MODEL, build_sae_cfg, load_model_cached
 
 
-@pytest.fixture
-def cfg():
-    cfg = build_sae_cfg(d_in=64, d_sae=128, hook_layer=0)
+@pytest.fixture(params=["standard", "gated"])
```
Comment (jettjaniak) [inline tests/unit/training/test_sae_trainer.py:22]: maybe too much grid search

File: tests/unit/training/test_sae_trainer.py [hunk 7]
```diff
@@ -185,6 +186,7 @@ def test_build_train_step_log_dict(trainer: SAETrainer) -> None:
         # l1 loss is scaled by l1_coefficient
         "losses/l1_loss": train_output.l1_loss / trainer.cfg.l1_coefficient,
         "losses/ghost_grad_loss": pytest.approx(0.15),
+        "losses/gated_aux_loss": pytest.approx(0.28),
```
Comment (jettjaniak) [inline tests/unit/training/test_sae_trainer.py:189]: don't need to be approx?
